計算コマンドに対応する

ochaochaocha3 opened the pull request:
これまで実行できなかった計算コマンド `C(...)` を実行できるようにしました。これにより、QuoridornなどのBCDice APIを使用するTRPGツールでも、どどんとふと同様に四則演算の結果を確認することができるようになります。

# 実行例

<img width="805" alt="bcdice-api" src="https://user-images.githubusercontent.com/2551173/61198736-4299b700-a716-11e9-83e4-15d4341efc48.png">

ochaochaocha3 committed:
Support calculation command

ochaochaocha3 committed:
Rakefile: Fix library directories in the test task

ysakasin commented on lib/bcdice_wrap.rb:
```diff
@@ -18,6 +18,20 @@ class BCDice
+    # 空白が含まれる場合、最初の部分だけを取り出す
+    target = s.split(/\s/, 2).first
+
+    return %r{\AC\([-+*/()\d]+\)\z}i === target
```
- splitせずに `%r{^C\([-+*/()\d]+\)\z}i` だけをテストすればよさそう
  - `s.split(/\s/)` の先頭がコマンドになっているということは、sは空白から始まっていない
- チェックには `Regexp#match?` を用いてください
  - https://docs.ruby-lang.org/ja/latest/method/Regexp/i/match=3f.html

ysakasin replied:
もしくは、もっと省略して `String#start_with?` を使ってもいいかもしれません

ochaochaocha3 replied:
> * splitせずに `%r{^C\([-+*/()\d]+\)\z}i` だけをテストすればよさそう
>     * `s.split(/\s/)` の先頭がコマンドになっているということは、sは空白から始まっていない

基本的にはBCDiceからそのまま持ってきたものです。BCDiceでは以下の `C(10+5) mokekeke` のような入力も通りますが、そのように変更すると通らなくならないでしょうか？

https://github.com/bcdice/BCDice/blob/f8881ea551565592580bb1dd9d8348f2d2cdfb58/src/test/data/None.txt#L229-L233

> * チェックには `Regexp#match?` を用いてください
>       * https://docs.ruby-lang.org/ja/latest/method/Regexp/i/match=3f.html

`Regexp#match?` はRuby 2.4の新機能だそうです。導入するとRuby 2.4以降のみの対応に変わりますが、よろしいでしょうか？

https://qiita.com/jnchito/items/9f9d45581816f121af07#%E6%AD%A3%E8%A6%8F%E8%A1%A8%E7%8F%BE%E3%81%AB%E6%96%87%E5%AD%97%E5%88%97%E3%81%8C%E3%83%9E%E3%83%83%E3%83%81%E3%81%97%E3%81%9F%E3%81%8B%E3%81%A9%E3%81%86%E3%81%8B%E3%81%A0%E3%81%91%E3%82%92%E8%BF%94%E3%81%99regexpmatchstringmatchsymbolmatch

ysakasin replied:
`\z`の仕様を勘違いしていました。 `%r{^C\([-+*/()\d]+\)}i` にすればよさそうです。

ysakasin replied:
Ruby 2.3はEOLなので、サポートから外そうと思います。`Regexp#match?` を用いて大丈夫です。

このままだとCIにRuby 2.3があることでテストが落ちると思いますので、CIの設定からRuby 2.3を消してもらえますか？

ochaochaocha3 replied:
コメントありがとうございます。以上のように修正いたします。

ochaochaocha3 replied:
すみません、BCDiceの挙動と合わせるためには、やはり `String#split` 相当の処理が必要でした。単純な前方マッチ `s.match?(%r{\AC\([-+*/()\d]+\)}i)` は次の両方の場合にマッチしますが、計算コマンドかどうかをBCDiceの挙動通りに正しく判定できませんでした。

* `C(10+5)mokeke`（間に半角空白なし）：BCDiceでは反応しない -> `seem_to_be_calc?` は `false` にしたい
* `C(10+5) mokeke`（間に半角空白あり）：BCDiceでは反応する -> `seem_to_be_calc?` は `true` にしたい

ysakasin replied:
それは把握していますが、広い方に互換性があるので、bcdice-apiとしては `C(10+5)mokeke` を許容してもいいと思います。

ochaochaocha3 replied:
分かりました。それでは、テストケースも含めて許容するように変えてみます。

ochaochaocha3 replied:
83d1e0154a3f1d3900216a8ac1f518e7805d3c15 で正規表現パターンを変更しました。

ysakasin commented on lib/bcdice_wrap.rb:
```diff
@@ -48,6 +62,24 @@ def dice_command   # ダイスコマンドの分岐処理
+    end
+
+    calc_result = matches[1]
+    return "#{@diceBot.gameType} : 計算結果 ＞ #{calc_result}", false
```
BCDice-APIでは結果の頭に `gameType` をつけていないので、この挙動は誤りです。

ysakasin replied:
`": 計算結果 ＞ #{calc_result}"`が期待される挙動です

ochaochaocha3 replied:
勘違いでした。修正いたします。

ysakasin commented on server.rb:
```diff
@@ -32,6 +32,19 @@ def diceroll(system, command)
+      # 数字のみが書かれているコマンドなどは拒絶するために必要な処理。
+      # BCDice側では、設定されたメッセージが計算コマンドかどうかの判定を行って
+      # いないため、やむを得ずここで判定する。
+      if BCDice.seem_to_be_calc?(command)
```
このチェックは `try_calc_command` に含めていいと思います。ネストを浅くできるのと、`try_calc_command` も結局は `nil, nil` を返す可能性があるので。

ochaochaocha3 replied:
そのように変更いたします。

ochaochaocha3 replied:
この変更は無理でした。`try_calc_command` では括弧内前処理が完了しているので、`@message` が `C123` のように `/C-?\d+/` の形になっており、判定できませんでした。

ysakasin replied:
`try_calc_command` が元のコマンドを引数で受け取れば良いと思っていたのですが、それだと見通しが悪いでしょうか？

ochaochaocha3 replied:
> `try_calc_command` が元のコマンドを引数で受け取れば良いと思っていたのですが、それだと見通しが悪いでしょうか？

これはできなさそうです。その理由は、BCDiceには計算コマンド `C(...)` 用の特別な処理が含まれていないからです。計算コマンドの実態は、括弧内前処理の結果として `C(...)` の括弧部分が勝手にただの整数に変換されて `C整数` となった後で、「整数」の部分を取り出す処理です。そのため、正規表現 `C(-?\d+)` が出てきます。元のコマンドを引数で受け取るのは確かにやりたいことなのですが、dengineやGoBCDiceで実装しているような構文解析器が要りそうです😇

`try_calc_command` はBCDiceの
https://github.com/bcdice/BCDice/blob/f8881ea551565592580bb1dd9d8348f2d2cdfb58/src/bcdiceCore.rb#L631-L637
に相当する処理で、とりあえずこれを移しておけば動くだろうという判断でした。

ysakasin replied:
```
    if result.nil?
      result, secret = bcdice.try_calc_command(command)
    end
```

という意図だったのですが、書けないのでしょうか？

ochaochaocha3 replied:
すみません、意図をきちんと読み取れていませんでした。以下のようにすれば確かにできそうです。

```ruby
class BCDice
  def try_calc_command(command)
    unless self.class.seem_to_be_calc?(command)
      return nil, nil
    end

    # 以下では @message を使って計算結果を返す
  end
end
```

ochaochaocha3 replied:
33a1ffe7c7310cce9d894a8c7f13b5091b476523 で上記のようにインターフェースを変更しました。

ysakasin commented on lib/bcdice_wrap.rb:
```diff
@@ -48,6 +62,24 @@ def dice_command   # ダイスコマンドの分岐処理
+  # 用意している。
+  def try_calc_command
+    stripped_message = @message.strip
+    matches = stripped_message.match(/\AC([-\d]+\z)/i)
```
より正確に `/\AC(-?\d+)\z/i` の方が良いかと思います。

ochaochaocha3 replied:
そのように変更いたします。

ochaochaocha3 committed:
Fix result message of calc command

refs #8

Game type should not be included in the result message.

ochaochaocha3 committed:
Fix regexps for calc command

refs #8

ochaochaocha3 committed:
Travis CI: Drop Ruby 2.3 (EOL)

refs #8

ochaochaocha3 committed:
API test: Add testcases

refs #8

ochaochaocha3 committed:
Expand the regex in BCDice.seem_to_be_calc?

refs #8

This commit changes BCDice.seem_to_be_calc? to return true for
pseudo-calc commands (e.g. "C(10+5)mokekeke") in which calc pattern
and the other words are concatenated without space.

ochaochaocha3 committed:
Move calc test into BCDice#try_calc_command

refs #8

ysakasin commented:
ありがとうございます。マージしました。

ochaochaocha3 commented:
こちらこそ、丁寧にレビューしていただき、ありがとうございました。
